File: LectureNotes/Unit01-Files/dieBattle.js
// --- Die Class ---
class Die {
    /** Represents a single, standard six-sided die. */
    roll() {
        // Returns a random integer between 1 and 6, inclusive.
        return Math.floor(Math.random() * 6) + 1;
    }
}

// --- Player Class ---
class Player {
    /** Represents a player with a name and a roll value. */
    constructor(name) {
        this.name = name;
        this.rollValue = 0;
    }

    /** Rolls the provided Die object and saves the result. */
    rollDie(die) {
        this.rollValue = die.roll();
        console.log(`${this.name} rolled a ${this.rollValue}.`);
    }
}

// --- DiceBattle Class ---
class DiceBattle {
    /** Manages the two players and the single die for the game. */
    constructor(p1Name, p2Name) {
        this.die = new Die();               // Composition: Game has a Die
        this.player1 = new Player(p1Name); // Composition: Game has Player 1
        this.player2 = new Player(p2Name); // Composition: Game has Player 2
    }

    /** Executes the game logic. */
    play() {
        console.log("\n--- Dice Battle Start (Node.js) ---");
        
        // Players take turns rolling the same die object
        this.player1.rollDie(this.die);
        this.player2.rollDie(this.die);

        // Determine the winner
        if (this.player1.rollValue > this.player2.rollValue) {
            console.log(`\n${this.player1.name} wins! (High Roll)`);
        } else if (this.player2.rollValue > this.player1.rollValue) {
            console.log(`\n${this.player2.name} wins! (High Roll)`);
        } else {
            console.log("\nIt's a draw!");
        }
        console.log("-----------------------------------");
    }
}

// --- Main Execution ---
// Initialize and play the game
const game = new DiceBattle("PlayerA", "PlayerB");
game.play();

File: LectureNotes/Unit01-Files/dieBattle.py
import random

class Die:
    """Represents a single, standard six-sided die."""
    def roll(self):
        # Returns a random integer between 1 and 6, inclusive.
        return random.randint(1, 6)

class Player:
    """Represents a player with a name and a roll value."""
    def __init__(self, name):
        self.name = name
        self.roll_value = 0

    def roll_die(self, die):
        """Rolls the provided Die object and saves the result."""
        self.roll_value = die.roll()
        print(f"{self.name} rolled a {self.roll_value}.")

class DiceBattle:
    """Manages the two players and the single die for the game."""
    def __init__(self, p1_name, p2_name):
        self.die = Die() # Composition: Game has a Die
        self.player1 = Player(p1_name) # Composition: Game has Player 1
        self.player2 = Player(p2_name) # Composition: Game has Player 2

    def play(self):
        """Executes the game logic."""
        print("\n--- Dice Battle Start ---")
        
        # Players take turns rolling the same die object
        self.player1.roll_die(self.die)
        self.player2.roll_die(self.die)

        # Determine the winner
        if self.player1.roll_value > self.player2.roll_value:
            winner = self.player1.name
            print(f"\n{winner} wins! (High Roll)")
        elif self.player2.roll_value > self.player1.roll_value:
            winner = self.player2.name
            print(f"\n{winner} wins! (High Roll)")
        else:
            print("\nIt's a draw!")

if __name__ == "__main__":
    # Initialize and play the game
    game = DiceBattle("Alice", "Bob")
    game.play()
